### PR TITLE
Feature/paint

### PR DESCRIPTION
**Pull request by HannahRKent**

### CODE REVIEW

File: src/main/java/de/orchound/gameoflife/model/Game.java [hunk 1]
```diff
@@ -96,6 +117,17 @@ public void randomize() {
 		notifyBoardDataObservers();
 	}
 
+	public void clear() {
+		Arrays.fill(board.target, false);
+		board.makeCurrentStateInitial();
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Game.java]: Move these to a `clear()` method in the `Board`.
```suggestion
		board.clear();
```

File: src/main/java/de/orchound/gameoflife/model/Game.java [hunk 2]
```diff
@@ -66,11 +69,29 @@ public void toggleCell(Vector2ic cell) {
...
-		else
+			paintMode = false;
+		}
+		else {
 			board.resurrectCell(cell.y(), cell.x());
+			paintMode = true;
+		}
+		notifyBoardDataObservers();
 
+
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Game.java]: Remove unnecessary line breaks.

File: src/main/java/de/orchound/gameoflife/model/Game.java [hunk 3]
```diff
@@ -66,11 +69,29 @@ public void toggleCell(Vector2ic cell) {
...
+	}
+
+	public void setCell(Vector2ic cell) {
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Game.java]: That's a pretty confusing / misleading name, since it suggests that we either set a cell to living or set it to some state provided as parameter.
The occurring state change of the cell depends on the "paint mode", so maybe we should call it `drawCell()`?

I still think it wouldn't be very clear to the client what this method ends up doing to the cell.
It looks like cell painting is so arbitrary, the determination of what cells to set or unset should be taken care of by another class.

`Game` is for representing and advancing the game state. It shouldn't implement the drawing logic.
You should rather add a function or two to set and unset individual cells and then extract the drawing logic.

File: src/main/java/de/orchound/gameoflife/model/Game.java [hunk 4]
```diff
@@ -66,11 +69,29 @@ public void toggleCell(Vector2ic cell) {
...
+		if (paintMode) {
+			if (!board.getCellStatus(cell.y(), cell.x())) {
+				board.resurrectCell(cell.y(), cell.x());
+			}
+		} else {
+			if (board.getCellStatus(cell.y(), cell.x())) {
+				board.killCell(cell.y(), cell.x());
+			}
+		}
 		notifyBoardDataObservers();
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Game.java]: Notifying the observers here is somewhat costly and should only be done if the the board state actually changed.
This is currently not the case. The function might end up not changing the state at all and still communicates an update to the observers.

File: src/main/java/de/orchound/gameoflife/model/Board.java [hunk 5]
```diff
@@ -59,11 +59,13 @@ public boolean getCellStatus(int rowIndex, int cellIndex) {
 	}
 
 	public void resurrectCell(int rowIndex, int cellIndex) {
-		target[rowIndex * size.x() + cellIndex] = true;
+		if (cellIndex >= 0 & cellIndex < size.x())
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Board.java]: `&` vs `&&`
We also don't have a safe-guard like this for `getCellStatus()`. I think since the board dimensions are known to the client, it should be the client's responsibility to make sure to pass the correct indices.
So move these guards to `Game`.

File: src/test/java/de/orchound/gameoflife/model/BoardTest.java [hunk 6]
```diff
@@ -18,11 +18,57 @@ void setUp() {
 	}
 
 	@Test
-	void testSetCellAlive() {
+	void testResurrectCellZeroCellIndex() {
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/model/BoardTest.java]: Please don't test each edge case in a dedicated method.
Simply implement `testResurrectCell()` and `testKillCell()`, then put as many assertions in as you might need to test the edge cases.

File: src/test/java/de/orchound/gameoflife/model/BoardTest.java [hunk 7]
```diff
@@ -18,11 +18,57 @@ void setUp() {
...
 	}
 
+	@Test
+	void testResurrectCellMaxCellIndex() {
+		board.resurrectCell(1, 19);
+		assertTrue(board.getCellStatus(1, 19));
+	}
+
+	@Test
+	void testResurrectCellNegativeCellIndex() {
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/model/BoardTest.java]: Since I consider it the `Board` clients responsibility to provide valid indices, we don't need test methods like this at all.

File: src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java [hunk 8]
```diff
@@ -99,17 +101,27 @@ private void drawHud() {
...
 		if (mouseButton == LEFT) {
 			viewOffset.add(mouseX - pmouseX, mouseY - pmouseY);
+		} else if (mouseButton == RIGHT) {
+			try {
+				Vector2f position = bufferVector2f.set(mouseX, mouseY)
+					.sub(viewOffset)
+					.div(scale);
+				Vector2i cell = boardView.getCellAt(position, bufferVector2i);
+				game.setCell(cell);
+			}  catch (ArrayIndexOutOfBoundsException ignored) {}
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java]: Don't catch and ignore these errors, prevent them from happening.
Just as done in `Game.toggleCell()`, do input validation in `Game.setCell()`.

**Review by kBlaszczyk — CHANGES_REQUESTED** [review]

**Review by kBlaszczyk — APPROVED** [review]
